dipy/denoise/shift_twist_convolution.pyx

```diff
@@ -0,0 +1,190 @@
+    lut : array of double
+        The 5D lookup table
+    test_mode : boolean
+        Reduced convolution in one direction only for testing
+        
+    Returns
+    -------
+    output : array of double
+        The ODF data after convolution enhancement
+    """
+        
+    cdef:
+        double [:, :, :, ::1] output = np.array(odfs, copy=True)
+        cnp.npy_intp OR1 = lut.shape[0]
+        cnp.npy_intp OR2 = lut.shape[1]
+        cnp.npy_intp N = lut.shape[2]
+        cnp.npy_intp hn = (N - 1) / 2
+        double [:, :, :, :] totalval
+        double [:, :, :, :] voxcount
+        cnp.npy_intp nx = odfs.shape[0]
+        cnp.npy_intp ny = odfs.shape[1]
+        cnp.npy_intp nz = odfs.shape[2]
+        cnp.npy_intp threads_to_use = -1
+        cnp.npy_intp all_cores = openmp.omp_get_num_procs()
+        cnp.npy_intp corient, orient, cx, cy, cz, x, y, z
+        cnp.npy_intp expectedvox
+        cnp.npy_intp edgeNormalization = True
+
+    if num_threads is not None:
+        threads_to_use = num_threads
+    else:
+        threads_to_use = all_cores
+
+    if have_openmp:
+        openmp.omp_set_dynamic(0)
+        openmp.omp_set_num_threads(threads_to_use)
+    
+    if test_mode:
+        edgeNormalization = False
+        OR2 = 1
+
+    # expected number of voxels in kernel
+    totalval = np.zeros((OR1, nx, ny, nz))
+    voxcount = np.zeros((OR1, nx, ny, nz))
+    expectedvox = nx * ny * nz
+
+    with nogil:
+
+        # loop over ODFs cx,cy,cz,corient --> y and v
+        for corient in prange(OR1, schedule='guided'):
+            for cx in range(nx):
+                for cy in range(ny):
+                    for cz in range(nz):
+                        # loop over kernel x,y,z,orient --> x and r
+                        for x in range(int_max(cx - hn, 0), 
+                                       int_min(cx + hn + 1, ny - 1)):
+                             for y in range(int_max(cy - hn, 0), 
+                                            int_min(cy + hn + 1, ny - 1)):
+                                 for z in range(int_max(cz - hn, 0), 
+                                                int_min(cz + hn + 1, nz - 1)):
+                                    voxcount[corient, cx, cy, cz] += 1.0
+                                    for orient in range(0, OR2):
+                                        totalval[corient, cx, cy, cz] += \
+                                            odfs[x, y, z, orient] * \
+                                            lut[corient, orient, x - (cx - hn), y - (cy - hn), z - (cz - hn)]
+                        if edgeNormalization:
+                            output[cx, cy, cz, corient] = \
+                                totalval[corient, cx, cy, cz] * expectedvox/voxcount[corient, cx, cy, cz]
+                        else:
+                            output[cx, cy, cz, corient] = \
+                                totalval[corient, cx, cy, cz]
+
+    if have_openmp and num_threads is not None:
```

Comment:
This resets the number of OpenMP cores used to default, I've added a comment here